File: leetcode/5048.FindWordsThatCanBeFormedbyCharacters/Solution.py
from collections import defaultdict
import copy


class Solution:
    def countCharacters(self, words, chars):
        res = 0
        hash_map = defaultdict(int)
        for c in chars:
            hash_map[c] += 1

        for w in words:
            m = copy.deepcopy(hash_map)
            for c in w:
                if c in m and m[c] > 0:
                    m[c] -= 1
                else:
                    break
            else:
                res += len(w)

        # print(res)
        return res


s = Solution()
words = ["hello", "world", "leetcode"]
chars = "welldonehoneyr"
s.countCharacters(words, chars)

File: leetcode/5048.FindWordsThatCanBeFormedbyCharacters/Solution.cpp
#include "../../include/global.hpp"
#include <bits/stdc++.h>

using namespace std;

class Solution {
  public:
    int countCharacters(vector<string>& words, string chars) {
        int res = 0;
        map<char, int> hash_map;
        for (char c : chars) hash_map[c]++;
        for (string s : words) {
            bool flag = true;
            map<char, int> m = hash_map;
            for (char cc : s) {
                m[cc]--;
                // XXX: 省去判断 m has cc 的麻烦
                if (m[cc] < 0) {
                    flag = false;
                    break;
                }
            }
            if (flag) res += s.size();
        }
        return res;
    }
};
